hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -706,7 +706,20 @@ protected void compactIfNecessary(AbstractHoodieWriteClient writeClient, String
     }
   }
 
-  protected void doClean(AbstractHoodieWriteClient writeClient, String instantTime) {
+  protected void cleanIfNecessary(AbstractHoodieWriteClient writeClient, String instantTime) {
+    Option<HoodieInstant> lastCompletedCompactionInstant = metadataMetaClient.reloadActiveTimeline()
+        .getCommitTimeline().filterCompletedInstants().lastInstant();
+    if (lastCompletedCompactionInstant.isPresent()
+        && metadataMetaClient.getActiveTimeline().filterCompletedInstants()
+            .findInstantsAfter(lastCompletedCompactionInstant.get().getTimestamp()).countInstants() < 3) {
+      // do not clean the log files immediately after compaction to give some buffer time for metadata table reader,
```

Comment:
Now the default compaction delta commits is 10, the min retain commits is 20, the max retain commits is 30.

When the second compaction schedules and triggers, there are about 22 commits on the timeline, which would then trigger the cleaning immediately.

Reply:
min and max commits is for archival. cleaner is controlled by hoodie.metadata.cleaner.commits.retained which is 3. But I am good with the change.

Reply:
So this problem should also exist in the MOR table data path? Is there any solution there?

Reply:
I guess so, we need some protection logic too for the MOR table log files or even COW table parquet files, because if we write COW table very frequently (like each 10 seconds), the reader would very probably encounter `FileNotFoundException` .